File: runtime/sdk/node/velix_process.js
const net = require('net');
const fs = require('fs');
const os = require('os');
const crypto = require('crypto');

class VelixProcess {
  constructor(name, role) {
    this.processName = name;
    this.role = role;
    this.osPid = process.pid;
    this.velixPid = -1;
    this.treeId = '';
    this.parentPid = Number(process.env.VELIX_PARENT_PID || -1);
    this.entryTraceId = process.env.VELIX_TRACE_ID || '';
    this.userId = process.env.VELIX_USER_ID || '';
    this.params = this.#parseJson(process.env.VELIX_PARAMS || '{}');
    this.isRunning = false;
    this.busSocket = null;
    this.responses = new Map();
    this.pendingToolWaiters = new Map();
    this.status = 'STARTING';
  }

  async run() {
    throw new Error('run() must be overridden');
  }

  async start() {
    if (this.isRunning) return;
    this.isRunning = true;

    const reg = {
      message_type: 'REGISTER_PID',
      payload: {
        register_intent: this.parentPid <= 0 ? 'NEW_TREE' : 'JOIN_PARENT_TREE',
        role: this.role,
        os_pid: this.osPid,
        process_name: this.processName,
        trace_id: this.entryTraceId,
        status: 'STARTING',
        memory_mb: 0,
      },
    };
    if (this.parentPid > 0) reg.source_pid = this.parentPid;

    const reply = await this.#request('SUPERVISOR', reg, 5000);
    if (reply.status !== 'ok') {
      throw new Error(`registration failed: ${JSON.stringify(reply)}`);
    }

    this.velixPid = Number(reply.process?.pid || -1);
    this.treeId = String(reply.process?.tree_id || '');

    await this.#connectBus();
    this.#startHeartbeat();

    try {
      await this.run();
    } finally {
      this.shutdown();
    }
  }

  shutdown() {
    this.isRunning = false;
    if (this.busSocket) {
      this.busSocket.destroy();
      this.busSocket = null;
    }
  }

  async callLlm(convoId, userMessage = '', systemMessage = '', userId = '', mode = 'user_conversation') {
    return this._callLlmInternal(convoId, userMessage, systemMessage, userId, mode);
  }

  /**
   * Resumes a conversation with an out-of-band tool result.
   */
  async callLlmResume(convoId, toolResult, userId = '') {
    return this._callLlmInternal(convoId || '', '', '', userId || '', 'user_conversation', toolResult);
  }

  async _callLlmInternal(convoId, userMessage, systemMessage, userId, mode, toolResultOverride = null) {
    const effectiveUserId = userId || '';
    const m = mode || (!convoId && !effectiveUserId ? 'simple' : (effectiveUserId ? 'user_conversation' : 'conversation'));

    const basePayload = {
      message_type: 'LLM_REQUEST',
      mode: m,
      tree_id: this.tree_id,
      source_pid: this.velixPid,
      priority: 1,
      convo_id: convoId || '',
      user_id: effectiveUserId,
      owner_pid: this.velixPid,
    };

    let nextMessages = [];
    if (systemMessage) nextMessages.push({ role: 'system', content: systemMessage });
    if (userMessage) nextMessages.push({ role: 'user', content: userMessage });

    const maxIterations = 15;
    for (let i = 0; i < maxIterations; i++) {
      this.status = 'WAITING_LLM';

      const payload = {
        ...basePayload,
        request_id: `req_${this.velixPid}_${crypto.randomUUID().slice(0, 8)}`,
        trace_id: crypto.randomUUID().replace(/-/g, ''),
      };

      if (i === 0 && toolResultOverride) {
        payload.tool_message = toolResultOverride;
      }

      if (nextMessages.length) {
        payload.messages = [...nextMessages];
      }

      const resp = await this.#request('LLM_SCHEDULER', payload, 120000);
      if (resp.status !== 'ok') {
        this.status = 'ERROR';
        throw new Error(resp.error || 'llm request failed');
      }

      if (!resp.tool_calls || !Array.isArray(resp.tool_calls) || resp.tool_calls.length === 0) {
        this.status = 'RUNNING';
        return String(resp.response || '');
      }

      const toolCalls = resp.tool_calls;
      const toolMessages = [];
      let toolExecuted = false;

      for (const toolCall of toolCalls) {
        const name = String(toolCall?.function?.name || toolCall?.name || '').trim();
        if (!name) continue;
        const args = toolCall?.function?.arguments || toolCall?.arguments || {};

        this.status = 'RUNNING';
        const result = await this.executeTool(name, args);
        toolExecuted = true;
        toolMessages.push({
          role: 'tool',
          content: JSON.stringify(result),
          tool_call_id: toolCall.id || toolCall.trace_id || '',
        });
      }

      if (!toolExecuted) {
        this.status = 'RUNNING';
        return String(resp.response || '');
      }

      nextMessages = toolMessages;
    }

    this.status = 'ERROR';
    return 'Failure: Agent state machine exceeded max iterations.';
  }

  #extractToolCalls(reply) {
    const out = [];

    if (Array.isArray(reply?.exec_blocks)) {
      for (const block of reply.exec_blocks) {
        if (block && typeof block === 'object' && !Array.isArray(block)) {
          out.push(block);
          continue;
        }
        if (typeof block === 'string') {
          const raw = block.trim();
          if (!raw) continue;
          try {
            const parsed = JSON.parse(raw);
            if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
              out.push(parsed);
            }
          } catch (_) {}
        }
      }
    }

    if (Array.isArray(reply?.tool_calls)) {
      for (const call of reply.tool_calls) {
        if (call && typeof call === 'object') {
          out.push(call);
        }
      }
    }

    const responseText = String(reply?.response || '');
    let searchAt = 0;
    while (true) {
      const start = responseText.indexOf('EXEC', searchAt);
      if (start === -1) break;
      const end = responseText.indexOf('END_EXEC', start);
      if (end === -1) break;

      const raw = responseText.slice(start + 4, end).trim();
      if (raw) {
        try {
          const parsed = JSON.parse(raw);
          if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
            out.push(parsed);
          }
        } catch (_) {}
      }
      searchAt = end + 'END_EXEC'.length;
    }

    return out;
  }

  async executeTool(name, params) {
    const traceId = crypto.randomUUID().replace(/-/g, '');
    const req = {
      message_type: 'EXEC_VELIX_PROCESS',
      trace_id: traceId,
      tree_id: this.treeId,
      source_pid: this.velixPid,
      name,
      params,
    };
    if (this.userId) {
      req.user_id = this.userId;
    }
    const ack = await this.#request('EXECUTIONER', req, 5000);
    if (ack.status !== 'ok') {
      throw new Error(ack.error || 'executioner rejected');
    }

    if (this.responses.has(traceId)) {
      const payload = this.responses.get(traceId);
      this.responses.delete(traceId);
      if (payload && payload.error === 'child_terminated') {
        throw new Error(`child terminated: ${JSON.stringify(payload)}`);
      }
      return payload || {};
    }

    return await new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        const waiters = this.pendingToolWaiters.get(traceId) || [];
        this.pendingToolWaiters.set(
          traceId,
          waiters.filter((w) => w.resolve !== resolve)
        );
        reject(new Error('tool result timeout'));
      }, 60 * 60 * 1000);

      const waiter = {
        resolve: (payload) => {
          clearTimeout(timeout);
          if (payload && payload.error === 'child_terminated') {
            reject(new Error(`child terminated: ${JSON.stringify(payload)}`));
            return;
          }
          resolve(payload || {});
        },
      };

      const waiters = this.pendingToolWaiters.get(traceId) || [];
      waiters.push(waiter);
      this.pendingToolWaiters.set(traceId, waiters);
    });
  }

  reportResult(targetPid, data, traceId = '', append = true) {
    if (!this.busSocket) return;
    const tid = traceId || this.entryTraceId;
    this.#sendFramed(this.busSocket, {
      message_type: 'IPM_RELAY',
      target_pid: targetPid,
      trace_id: tid,
      payload: data,
    });

    if (append) {
      // Logic for internal tool-calling wait-state
      // (Wait, Node.js uses responses/waiters map directly in executeTool)
    } else {
      // Async: remove trace from waiters so executeTool stops waiting.
      if (tid) {
        this.pendingToolWaiters.delete(tid);
        this.responses.delete(tid);
      }
    }
  }

  async #connectBus() {
    const port = this.#getPort('BUS', 5174);
    this.busSocket = await this.#connect('127.0.0.1', port, 5000);
    await this.#sendAndRecv(this.busSocket, { message_type: 'BUS_REGISTER', pid: this.velixPid });

    let buffer = Buffer.alloc(0);
    this.busSocket.on('data', (chunk) => {
      buffer = Buffer.concat([buffer, chunk]);
      while (buffer.length >= 4) {
        const size = buffer.readUInt32BE(0);
        if (buffer.length < 4 + size) break;
        const body = buffer.slice(4, 4 + size);
        buffer = buffer.slice(4 + size);
        try {
          const msg = JSON.parse(body.toString('utf-8'));
          if (msg.message_type === 'IPM_PUSH' || msg.message_type === 'CHILD_TERMINATED') {
            const traceId = msg.trace_id || '';
            if (traceId) {
              const payload = msg.payload || {};
              const waiters = this.pendingToolWaiters.get(traceId);
              if (waiters && waiters.length) {
                this.pendingToolWaiters.delete(traceId);
                for (const w of waiters) {
                  try {
                    w.resolve(payload);
                  } catch (_) {}
                }
              } else {
                this.responses.set(traceId, payload);
              }
            }
          }
        } catch (_) {}
      }
    });
  }

  #startHeartbeat() {
    const tick = async () => {
      if (!this.isRunning) return;
      try {
        await this.#request('SUPERVISOR', {
          message_type: 'HEARTBEAT',
          pid: this.velixPid,
          payload: { status: 'RUNNING', memory_mb: 0 },
        }, 3000);
      } catch (_) {}
      setTimeout(tick, 5000);
    };
    setTimeout(tick, 5000);
  }

  async #request(service, payload, timeoutMs) {
    const port = this.#getPort(service, 0);
    if (!port) throw new Error(`invalid port for ${service}`);
    const sock = await this.#connect('127.0.0.1', port, timeoutMs);
    try {
      return await this.#sendAndRecv(sock, payload);
    } finally {
      sock.destroy();
    }
  }

  #connect(host, port, timeoutMs) {
    return new Promise((resolve, reject) => {
      const sock = net.createConnection({ host, port });
      sock.setTimeout(timeoutMs);
      sock.once('connect', () => resolve(sock));
      sock.once('timeout', () => reject(new Error('socket timeout')));
      sock.once('error', reject);
    });
  }

  #sendAndRecv(sock, payload) {
    return new Promise((resolve, reject) => {
      let buffer = Buffer.alloc(0);
      let done = false;
      const onData = (chunk) => {
        if (done) return;
        buffer = Buffer.concat([buffer, chunk]);
        if (buffer.length < 4) return;
        const size = buffer.readUInt32BE(0);
        if (buffer.length < 4 + size) return;
        const body = buffer.slice(4, 4 + size);
        done = true;
        cleanup();
        try {
          resolve(JSON.parse(body.toString('utf-8')));
        } catch (e) {
          reject(e);
        }
      };
      const onError = (e) => {
        if (done) return;
        done = true;
        cleanup();
        reject(e);
      };
      const onTimeout = () => {
        if (done) return;
        done = true;
        cleanup();
        sock.destroy();
        reject(new Error('socket timeout'));
      };
      const cleanup = () => {
        sock.off('data', onData);
        sock.off('error', onError);
        sock.off('timeout', onTimeout);
      };
      sock.on('data', onData);
      sock.on('error', onError);
      sock.on('timeout', onTimeout);
      this.#sendFramed(sock, payload);
    });
  }

  #sendFramed(sock, payload) {
    const body = Buffer.from(JSON.stringify(payload), 'utf-8');
    const header = Buffer.alloc(4);
    header.writeUInt32BE(body.length, 0);
    sock.write(Buffer.concat([header, body]));
  }

  #getPort(name, fallback) {
    const candidates = ['config/ports.json', '../config/ports.json', 'build/config/ports.json'];
    try {
      for (const p of candidates) {
        if (!fs.existsSync(p)) continue;
        const raw = fs.readFileSync(p, 'utf-8');
        const ports = JSON.parse(raw);
        return Number(ports[name] || fallback);
      }
      return fallback;
    } catch (_) {
      return fallback;
    }
  }

  #parseJson(raw) {
    try {
      const obj = JSON.parse(raw);
      return typeof obj === 'object' && obj ? obj : {};
    } catch (_) {
      return {};
    }
  }
}

module.exports = { VelixProcess };

File: runtime/sdk/rust/velix_process.rs
use std::collections::HashMap;
use std::env;
use std::io::{Read, Write};
use std::net::Shutdown;
use std::net::TcpStream;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Condvar, Mutex};
use std::thread;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use serde_json::{json, Map, Value};

pub struct VelixProcess {
    pub process_name: String,
    pub role: String,
    pub os_pid: i32,
    pub velix_pid: i32,
    pub tree_id: String,
    pub parent_pid: i32,
    pub entry_trace_id: String,
    pub user_id: String,
    pub params: Value,
    status: Arc<Mutex<String>>,
    is_running: Arc<AtomicBool>,
    bus_stream: Option<TcpStream>,
    responses: Arc<(Mutex<HashMap<String, Value>>, Condvar)>,
}

impl VelixProcess {
    pub fn new(name: &str, role: &str) -> Self {
        Self {
            process_name: name.to_string(),
            role: role.to_string(),
            os_pid: std::process::id() as i32,
            velix_pid: -1,
            tree_id: String::new(),
            parent_pid: env::var("VELIX_PARENT_PID").ok().and_then(|v| v.parse().ok()).unwrap_or(-1),
            entry_trace_id: env::var("VELIX_TRACE_ID").unwrap_or_default(),
            user_id: env::var("VELIX_USER_ID").unwrap_or_default(),
            params: parse_env_json("VELIX_PARAMS"),
            status: Arc::new(Mutex::new("STARTING".to_string())),
            is_running: Arc::new(AtomicBool::new(false)),
            bus_stream: None,
            responses: Arc::new((Mutex::new(HashMap::new()), Condvar::new())),
        }
    }

    pub fn start<F>(&mut self, mut run: F) -> Result<(), String>
    where
        F: FnMut(&mut VelixProcess) -> Result<(), String>,
    {
        if self.is_running.load(Ordering::SeqCst) {
            return Ok(());
        }
        self.is_running.store(true, Ordering::SeqCst);

        let mut reg = json!({
            "message_type": "REGISTER_PID",
            "payload": {
                "register_intent": if self.parent_pid <= 0 { "NEW_TREE" } else { "JOIN_PARENT_TREE" },
                "role": self.role,
                "os_pid": self.os_pid,
                "process_name": self.process_name,
                "trace_id": self.entry_trace_id,
                "status": "STARTING",
                "memory_mb": 0
            }
        });
        if self.parent_pid > 0 {
            reg["source_pid"] = json!(self.parent_pid);
        }

        let reply = self.request("SUPERVISOR", &reg, Duration::from_secs(5))?;
        if reply["status"] != "ok" {
            return Err(format!("registration failed: {}", reply));
        }
        self.velix_pid = reply["process"]["pid"].as_i64().unwrap_or(-1) as i32;
        self.tree_id = reply["process"]["tree_id"].as_str().unwrap_or("").to_string();

        self.connect_bus()?;
        self.start_heartbeat();

        let result = run(self);
        self.shutdown();
        result
    }

    pub fn shutdown(&mut self) {
        self.is_running.store(false, Ordering::SeqCst);
        if let Some(stream) = self.bus_stream.as_mut() {
            let _ = stream.shutdown(Shutdown::Both);
        }
        self.bus_stream = None;
    }

    pub fn call_llm(&self, convo_id: &str, user_message: &str, system_message: &str, user_id: &str, mode: &str) -> Result<String, String> {
        self.call_llm_internal(convo_id, user_message, system_message, user_id, mode, None)
    }

    pub fn call_llm_resume(&self, convo_id: &str, tool_result: Value, user_id: &str) -> Result<String, String> {
        self.call_llm_internal(convo_id, "", "", user_id, "user_conversation", Some(tool_result))
    }

    fn call_llm_internal(&self, convo_id: &str, user_message: &str, system_message: &str, user_id: &str, mode: &str, tool_result_override: Option<Value>) -> Result<String, String> {
        let m = if !mode.is_empty() {
            mode.to_string()
        } else if convo_id.is_empty() && user_id.is_empty() {
            "simple".to_string()
        } else if !user_id.is_empty() {
            "user_conversation".to_string()
        } else {
            "conversation".to_string()
        };

        let base_payload = json!({
            "message_type": "LLM_REQUEST",
            "mode": m,
            "tree_id": self.tree_id,
            "source_pid": self.velix_pid,
            "priority": 1,
            "convo_id": convo_id,
            "user_id": user_id,
            "owner_pid": self.velix_pid
        });

        let mut next_messages: Vec<Value> = Vec::new();
        if !system_message.is_empty() {
            next_messages.push(json!({"role": "system", "content": system_message}));
        }
        if !user_message.is_empty() {
            next_messages.push(json!({"role": "user", "content": user_message}));
        }

        const MAX_ITERATIONS: usize = 15;
        for i in 0..MAX_ITERATIONS {
            self.set_status("WAITING_LLM");

            let mut payload = base_payload.clone();
            payload["request_id"] = Value::String(format!("req_{}_{}", self.velix_pid, rand_suffix()));
            payload["trace_id"] = Value::String(format!("trace_{}", now_nanos()));

            if i == 0 {
                if let Some(tr) = &tool_result_override {
                    payload["tool_message"] = tr.clone();
                }
            }

            if !next_messages.is_empty() {
                payload["messages"] = Value::Array(next_messages.clone());
            }

            let resp = self.request("LLM_SCHEDULER", &payload, Duration::from_secs(120))?;
            if resp["status"] != "ok" {
                self.set_status("ERROR");
                return Err(resp["error"].as_str().unwrap_or("llm request failed").to_string());
            }

            let tool_calls = resp.get("tool_calls").and_then(|v| v.as_array());
            if tool_calls.is_none() || tool_calls.unwrap().is_empty() {
                self.set_status("RUNNING");
                return Ok(resp["response"].as_str().unwrap_or("").to_string());
            }

            let tool_calls = tool_calls.unwrap();
            let mut tool_messages: Vec<Value> = Vec::new();
            let mut tool_executed = false;

            for call in tool_calls {
                let name = call.get("function").and_then(|f| f.get("name")).and_then(|v| v.as_str())
                    .or_else(|| call.get("name").and_then(|v| v.as_str()))
                    .unwrap_or("")
                    .trim()
                    .to_string();
                if name.is_empty() {
                    continue;
                }

                let args = call.get("function").and_then(|f| f.get("arguments")).cloned()
                    .or_else(|| call.get("arguments").cloned())
                    .filter(|v| v.is_object())
                    .unwrap_or_else(|| Value::Object(Map::new()));

                self.set_status("RUNNING");
                let result = self.execute_tool(&name, args)?;
                tool_executed = true;

                let call_id = call.get("id").and_then(|v| v.as_str())
                    .or_else(|| call.get("trace_id").and_then(|v| v.as_str()))
                    .unwrap_or("");

                tool_messages.push(json!({
                    "role": "tool",
                    "content": result.to_string(),
                    "tool_call_id": call_id
                }));
            }

            if !tool_executed {
                self.set_status("RUNNING");
                return Ok(resp["response"].as_str().unwrap_or("").to_string());
            }

            next_messages = tool_messages;
        }

        self.set_status("ERROR");
        Ok("Failure: Agent state machine exceeded max iterations.".to_string())
    }

    pub fn execute_tool(&self, name: &str, params: Value) -> Result<Value, String> {
        let trace_id = format!("trace_{}", now_nanos());
        let req = json!({
            "message_type": "EXEC_VELIX_PROCESS",
            "trace_id": trace_id,
            "tree_id": self.tree_id,
            "source_pid": self.velix_pid,
            "name": name,
            "params": params
        });
        let mut req = req;
        if !self.user_id.is_empty() {
            req["user_id"] = Value::String(self.user_id.clone());
        }
        let ack = self.request("EXECUTIONER", &req, Duration::from_secs(5))?;
        if ack["status"] != "ok" {
            return Err(format!("executioner rejected: {}", ack));
        }

        let (lock, cvar) = &*self.responses;
        let mut map = lock.lock().map_err(|_| "response lock poisoned".to_string())?;
        let timeout_at = std::time::Instant::now() + Duration::from_secs(60 * 60);
        loop {
            if let Some(payload) = map.remove(&trace_id) {
                if payload["error"] == "child_terminated" {
                    return Err(format!("child terminated: {}", payload));
                }
                return Ok(payload);
            }
            if std::time::Instant::now() > timeout_at {
                return Err("tool result timeout".to_string());
            }
            let wait = cvar.wait_timeout(map, Duration::from_millis(100)).map_err(|_| "wait failed".to_string())?;
            map = wait.0;
        }
    }

    pub fn report_result(&mut self, target_pid: i32, data: Value, trace_id: &str, append: bool) -> Result<(), String> {
        if let Some(stream) = self.bus_stream.as_mut() {
            let tid = if trace_id.is_empty() { &self.entry_trace_id } else { trace_id };

            let msg = json!({
                "message_type": "IPM_RELAY",
                "target_pid": target_pid,
                "trace_id": tid,
                "payload": data
            });
            send_framed(stream, &msg).map_err(|e| e.to_string())?;

            if !append && !tid.is_empty() {
                let (lock, cvar) = &*self.responses;
                if let Ok(mut map) = lock.lock() {
                    map.remove(tid);
                    cvar.notify_all();
                }
            }
        }
        Ok(())
    }

    fn connect_bus(&mut self) -> Result<(), String> {
        let mut stream = connect("127.0.0.1", get_port("BUS", 5174), Duration::from_secs(5))?;
        let reg = json!({"message_type":"BUS_REGISTER", "pid": self.velix_pid});
        send_framed(&mut stream, &reg).map_err(|e| e.to_string())?;
        let _ = recv_framed(&mut stream).map_err(|e| e.to_string())?;

        let mut listener = stream.try_clone().map_err(|e| e.to_string())?;
        let responses = Arc::clone(&self.responses);
        let is_running = Arc::clone(&self.is_running);
        thread::spawn(move || {
            while is_running.load(Ordering::SeqCst) {
                let msg = match recv_framed(&mut listener) {
                    Ok(m) => m,
                    Err(_) => break,
                };
                let mt = msg["message_type"].as_str().unwrap_or("");
                if mt == "IPM_PUSH" || mt == "CHILD_TERMINATED" {
                    if let Some(trace_id) = msg["trace_id"].as_str() {
                        let (lock, cvar) = &*responses;
                        if let Ok(mut map) = lock.lock() {
                            map.insert(trace_id.to_string(), msg["payload"].clone());
                            cvar.notify_all();
                        }
                    }
                }
            }
        });

        self.bus_stream = Some(stream);
        Ok(())
    }

    fn start_heartbeat(&self) {
        let pid = self.velix_pid;
        let is_running = Arc::clone(&self.is_running);
        thread::spawn(move || {
            while is_running.load(Ordering::SeqCst) {
                let hb = json!({
                    "message_type": "HEARTBEAT",
                    "pid": pid,
                    "payload": {"status":"RUNNING", "memory_mb":0}
                });
                let _ = request_static("SUPERVISOR", &hb, Duration::from_secs(3));
                thread::sleep(Duration::from_secs(5));
            }
        });
    }

    fn request(&self, service: &str, payload: &Value, timeout: Duration) -> Result<Value, String> {
        request_static(service, payload, timeout)
    }

    fn set_status(&self, status: &str) {
        if let Ok(mut guard) = self.status.lock() {
            *guard = status.to_string();
        }
    }
}

fn extract_tool_calls(reply: &Value) -> Vec<Value> {
    let mut out: Vec<Value> = Vec::new();

    if let Some(arr) = reply.get("exec_blocks").and_then(|v| v.as_array()) {
        for item in arr {
            if item.is_object() {
                out.push(item.clone());
                continue;
            }
            if let Some(raw) = item.as_str() {
                let trimmed = raw.trim();
                if !trimmed.is_empty() {
                    if let Ok(parsed) = serde_json::from_str::<Value>(trimmed) {
                        if parsed.is_object() {
                            out.push(parsed);
                        }
                    }
                }
            }
        }
    }

    if let Some(arr) = reply.get("tool_calls").and_then(|v| v.as_array()) {
        for item in arr {
            if item.is_object() {
                out.push(item.clone());
            }
        }
    }

    let response_text = reply
        .get("response")
        .and_then(|v| v.as_str())
        .unwrap_or("");

    let mut search_at: usize = 0;
    while search_at < response_text.len() {
        let rel_start = match response_text[search_at..].find("EXEC") {
            Some(v) => v,
            None => break,
        };
        let start = search_at + rel_start;
        let rel_end = match response_text[start..].find("END_EXEC") {
            Some(v) => v,
            None => break,
        };
        let end = start + rel_end;

        let raw = response_text[start + "EXEC".len()..end].trim();
        if !raw.is_empty() {
            if let Ok(parsed) = serde_json::from_str::<Value>(raw) {
                if parsed.is_object() {
                    out.push(parsed);
                }
            }
        }

        search_at = end + "END_EXEC".len();
    }

    out
}

fn request_static(service: &str, payload: &Value, timeout: Duration) -> Result<Value, String> {
    let mut stream = connect("127.0.0.1", get_port(service, 0), timeout)?;
    send_framed(&mut stream, payload).map_err(|e| e.to_string())?;
    recv_framed(&mut stream).map_err(|e| e.to_string())
}

fn connect(host: &str, port: i32, timeout: Duration) -> Result<TcpStream, String> {
    if port <= 0 {
        return Err("invalid port".to_string());
    }
    TcpStream::connect_timeout(&(format!("{}:{}", host, port).parse().map_err(|_| "bad addr")?), timeout)
        .map_err(|e| e.to_string())
}

fn send_framed(stream: &mut TcpStream, payload: &Value) -> std::io::Result<()> {
    let body = serde_json::to_vec(payload)?;
    let size = (body.len() as u32).to_be_bytes();
    stream.write_all(&size)?;
    stream.write_all(&body)?;
    Ok(())
}

fn recv_framed(stream: &mut TcpStream) -> std::io::Result<Value> {
    let mut hdr = [0u8; 4];
    stream.read_exact(&mut hdr)?;
    let size = u32::from_be_bytes(hdr) as usize;
    let mut body = vec![0u8; size];
    stream.read_exact(&mut body)?;
    serde_json::from_slice(&body)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
}

fn get_port(name: &str, fallback: i32) -> i32 {
    for p in ["config/ports.json", "../config/ports.json", "build/config/ports.json"] {
        let raw = match std::fs::read_to_string(p) {
            Ok(v) => v,
            Err(_) => continue,
        };
        let v: Value = match serde_json::from_str(&raw) {
            Ok(v) => v,
            Err(_) => continue,
        };
        return v.get(name).and_then(|x| x.as_i64()).unwrap_or(fallback as i64) as i32;
    }
    fallback
}

fn parse_env_json(name: &str) -> Value {
    let raw = env::var(name).unwrap_or_default();
    if raw.is_empty() {
        return Value::Object(Map::new());
    }
    serde_json::from_str(&raw).unwrap_or(Value::Object(Map::new()))
}

fn now_nanos() -> u128 {
    SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or(Duration::from_secs(0)).as_nanos()
}

fn rand_suffix() -> u128 {
    now_nanos() % 100000000
}
